File: src/main/scala/com/michalplachta/galactic/db/DbClient.scala
package com.michalplachta.galactic.db

import com.michalplachta.galactic.values.{ Citizen, Tweet }

import scala.concurrent.ExecutionContext.Implicits.global
import scala.concurrent.Future

/**
 * A simulation of a client for a very slow database. Some calls may also fail with an exception ;)
 */
object DbClient {
  import com.michalplachta.galactic.db.FakeData._

  def findCitizenByName(citizenName: String): Future[Citizen] = {
    citizens.find(_.name == citizenName) match {
      case Some(citizen) ⇒ simulateResponse(citizen)
      case None          ⇒ simulateBadRequest(s"citizen with name $citizenName couldn't be found")
    }
  }

  def getFollowers(citizen: Citizen): Future[List[Citizen]] = {
    val followers = citizen.name match {
      case "Darth Vader"    ⇒ siths ++ clones
      case "Luke Skywalker" ⇒ jedis ++ rebels
      case _                ⇒ citizens.slice(citizens.indexOf(citizen), citizens.size)
    }
    simulateResponse(followers)
  }

  def getTweetsFor(citizen: Citizen): Future[List[Tweet]] = {
    val tweetsForCitizen = tweets.filter(_.author != citizen)
    simulateResponse(tweetsForCitizen)
  }

  private def simulateResponse[R](result: R): Future[R] = {
    Future {
      Thread.sleep(4000)
      result
    }
  }

  private def simulateBadRequest[R](error: String): Future[R] = {
    Future {
      Thread.sleep(2000)
      throw new Exception(s"Bad Request: $error")
    }
  }
}

File: src/main/java/com/michalplachta/galactic/java/db/DbClient.java
package com.michalplachta.galactic.java.db;

import com.michalplachta.galactic.java.values.Citizen;
import com.michalplachta.galactic.java.values.Tweet;
import javaslang.collection.List;
import javaslang.concurrent.Future;
import javaslang.concurrent.Promise;

import static com.michalplachta.galactic.java.db.FakeData.*;
import static javaslang.API.*;

/**
 * A simulation of a client for a very slow database. Some calls may also fail with an exception ;)
 */
public class DbClient {
    public static Future<? extends Citizen> getCitizenByName(String citizenName) {
        return citizens.find(citizen -> citizen.name.equals(citizenName))
                .map(DbClient::simulateResponse)
                .getOrElse(simulateBadRequest(String.format("citizen with name %s couldn't be found", citizenName)));
    }

    public static Future<List<? extends Citizen>> getFollowers(Citizen citizen) {
        List<Citizen> followers = Match(citizen.name).of(
                Case($("Darth Vader"), () -> siths.appendAll(clones)),
                Case($("Luke Skywalker"), () -> jedis.appendAll(rebels)),
                Case($(), () -> citizens.slice(citizens.indexOf(citizen), citizens.length()))
        );
        return simulateResponse(followers);
    }

    public static Future<List<Tweet>> getTweetsFor(Citizen citizen) {
        List<Tweet> tweetsForCitizen = tweets.filter(t -> !t.author.equals(citizen));
        return simulateResponse(tweetsForCitizen);
    }

    private static <R> Future<R> simulateResponse(R result) {
        Promise<R> promise = Promise.make();
        Future.run(() -> {
            Thread.sleep(4000);
            promise.success(result);
        });
        return promise.future();
    }

    private static <R> Future<R> simulateBadRequest(String error) {
        Promise<R> promise = Promise.make();
        Future.run(() -> {
            Thread.sleep(2000);
            promise.failure(new Exception("Bad Request: " + error));
        });
        return promise.future();
    }
}
